src/Renderer.cpp

```diff
@@ -375,6 +410,98 @@ void Renderer::ssao_pass()
 
 // --------------------------
 
+GLuint Renderer::upload_filter(filter_type ft)
+{
+    GLfloat gaussian[5] = {1, 4, 6, 4, 1};
+    GLfloat uniform[5] = {1, 1, 1, 1, 1};
+    GLfloat gaussian_big[11] = {0.090154, 0.090606, 0.090959, 0.091212, 0.091364, 0.091414, 0.091364, 0.091212, 0.090959, 0.090606, 0.090154};
+
+    switch (ft) {
+    case GAUSSIAN_RED_5:
+        glUseProgram(shaders[BLUR_RED_5]);
+        glUniform1fv(glGetUniformLocation(shaders[BLUR_RED_5], "kernel"), 5, gaussian);
+        glUniform1f(glGetUniformLocation(shaders[BLUR_RED_5], "magnitude"), 16);
+        return shaders[BLUR_RED_5];
+        break;
+    case GAUSSIAN_RGB_5:
+        glUseProgram(shaders[BLUR_RGB_5]);
+        glUniform1fv(glGetUniformLocation(shaders[BLUR_RGB_5], "kernel"), 5, gaussian);
+        glUniform1f(glGetUniformLocation(shaders[BLUR_RGB_5], "magnitude"), 16);
+        return shaders[BLUR_RGB_5];
+        break;
+    case GAUSSIAN_RGB_11:
+        glUseProgram(shaders[BLUR_RGB_11]);
+        glUniform1fv(glGetUniformLocation(shaders[BLUR_RGB_11], "kernel"), 11, gaussian_big);
+        glUniform1f(glGetUniformLocation(shaders[BLUR_RGB_11], "magnitude"), 1.0f);
+        return shaders[BLUR_RGB_11];
+        break;
+    case UNIFORM_RED_5:
+        glUseProgram(shaders[BLUR_RED_5]);
+        glUniform1fv(glGetUniformLocation(shaders[BLUR_RED_5], "kernel"), 5, uniform);
+        glUniform1f(glGetUniformLocation(shaders[BLUR_RED_5], "magnitude"), 5);
+        return shaders[BLUR_RED_5];
+        break;
+    case UNIFORM_RGB_5:
+        glUseProgram(shaders[BLUR_RGB_5]);
+        glUniform1fv(glGetUniformLocation(shaders[BLUR_RGB_5], "kernel"), 5, uniform);
+        glUniform1f(glGetUniformLocation(shaders[BLUR_RGB_5], "magnitude"), 5);
+        return shaders[BLUR_RGB_5];
+        break;
+    }
+}
+
+// --------------------------
+
+void Renderer::filter_pass(GLuint source_tex, GLuint target_fbo)
+{
+    glBindFramebuffer(GL_FRAMEBUFFER, target_fbo);
+    glClear(GL_COLOR_BUFFER_BIT);
+    glActiveTexture(GL_TEXTURE0);
+    glBindTexture(GL_TEXTURE_2D, source_tex);
+    glBindVertexArray(quad_vao);
+    glDrawArrays(GL_TRIANGLE_STRIP, 0, 4);
+}
+
+// --------------------------
+
+void Renderer::blur_red_texture(GLuint source_tex, GLuint fbo_tex, GLuint target_fbo, filter_type ft, int iterations)
+{
+    GLuint shader = upload_filter(ft);
+    glUseProgram(shader);
+
+    for (int i=0; i<iterations; i++) {
+        glUniform1i(glGetUniformLocation(shader, "horizontal"), true);
+        filter_pass(source_tex, ping_pong_fbo_red);
+        source_tex = fbo_tex;
+        glUniform1i(glGetUniformLocation(shader, "horizontal"), false);
```

Comment:
This should be done using two different shaders, one for horizontal and one for vertical. 

As it is right now we get branching in each of them, making calculations for both branches in each. We will improve performance by separating them into two different shaders.